ui/src/app/pages/analysis/runtime-panel.tsx

```diff
@@ -40,97 +40,171 @@ const styles = reactStyles({
+          fp.map('memory'),
+          // See above comment on CPU union.
+          fp.union([memory]),
+          fp.sortBy(fp.identity)
+        )(validLeonardoMachineTypes)}
+        onChange={
+          ({value}) => fp.flow(
+            fp.find({cpu, memory: value}),
+            // If the selected machine is not different from the current machine return null
+            maybeGetMachine,
+            onChange
+            )(validLeonardoMachineTypes) }
+        value={memory}
+        />
   </Fragment>;
 };
 
-const DiskSizeSelection = ({onChange, updatedDiskSize, masterDiskSize}) => {
-  return <div>
-    <label htmlFor='runtime-disk'
-          style={{marginRight: '.25rem'}}>Disk (GB)</label>
-      <InputNumber id='runtime-disk'
-                showButtons
-                decrementButtonClassName='p-button-secondary'
-                incrementButtonClassName='p-button-secondary'
-                value={updatedDiskSize || masterDiskSize}
-                inputStyle={{padding: '.75rem .5rem', width: '2rem'}}
-                onChange={({value}) => onChange(value === masterDiskSize ? null : value)}
-                min={50 /* Runtime API has a minimum 50GB requirement. */}/>
-  </div>;
+const DiskSizeSelection = ({onChange, updatedDiskSize, diskSize}) => {
+  return <Fragment>
+    <label htmlFor='runtime-disk' style={{marginRight: '.25rem'}}>Disk (GB)</label>
+    <InputNumber id='runtime-disk'
+      showButtons
+      decrementButtonClassName='p-button-secondary'
+      incrementButtonClassName='p-button-secondary'
+      value={updatedDiskSize || diskSize}
+      inputStyle={{padding: '.75rem .5rem', width: '2rem'}}
+      onChange={({value}) => onChange(value)}
+      min={50 /* Runtime API has a minimum 50GB requirement. */}/>
+  </Fragment>;
+};
+
+const DataProcConfig = ({onChange, dataprocConfig}) => {
+  const {
+    workerMachineType = defaultMachineType,
+    workerDiskSize = 50,
+    numberOfWorkers = 2,
+    numberOfPreemptibleWorkers = 0
+  } = dataprocConfig || {};
+  const [updatedNumWorkers, setUpdatedNumWorkers] = useState(numberOfWorkers);
+  const [updatedPreemtible, setUpdatedPreemptible] = useState(numberOfPreemptibleWorkers);
+  const [updatedWorkerMachine, setUpdatedWorkerMachine] = useState(null);
+  const [updatedDiskSize, setUpdatedDiskSize] = useState(workerDiskSize);
+
+  useEffect(() => {
+    const machineType = updatedWorkerMachine && updatedWorkerMachine.name;
+    const dataprocConfigChanged = updatedNumWorkers !== numberOfWorkers ||
+    updatedPreemtible !== numberOfPreemptibleWorkers ||
+    updatedDiskSize !== workerDiskSize ||
+    updatedWorkerMachine ||
+    !dataprocConfig;
+
+    onChange(dataprocConfigChanged ? {
+      workerMachineType: machineType,
+      workerDiskSize: updatedDiskSize,
+      numberOfWorkers: updatedNumWorkers,
+      numberOfPreemptibleWorkers: updatedPreemtible
+    } : null);
+
+    return () => onChange(null);
```

Comment:
Can you explain this - possibly via code comment? Why do we want to dispatch an onChange(null) on cleanup? Does this handle the scenario where the user switches the cloud service back to GCE?